src/coreclr/jit/importer_vectorization.cpp

```diff
@@ -0,0 +1,741 @@
+        GenTree* secondIndir = impCreateCompareInd(this, gtClone(data), TYP_INT, dataOffset + 2, value2);
+
+        // TODO-Unroll-CQ: Consider merging two indirs via XOR instead of QMARK
+        // e.g. gtNewOperNode(GT_XOR, TYP_INT, firstIndir, secondIndir);
+        // but it currently has CQ issues (redundant movs)
+        GenTreeColon* doubleIndirColon = gtNewColonNode(TYP_INT, secondIndir, gtNewFalse());
+        return gtNewQmarkNode(TYP_INT, firstIndir, doubleIndirColon);
+    }
+
+    assert(len >= 4 && len <= 8);
+
+    UINT64 value1 = MAKEINT64(cns[0], cns[1], cns[2], cns[3]);
+    if (len == 4)
+    {
+        //   [ ch1 ][ ch2 ][ ch3 ][ ch4 ]
+        //   [          value           ]
+        //
+        return impCreateCompareInd(this, data, TYP_LONG, dataOffset, value1);
+    }
+
+    // For 5..7 value2 will overlap with value1, e.g. for Length == 6:
+    //
+    //   [ ch1 ][ ch2 ][ ch3 ][ ch4 ][ ch5 ][ ch6 ]
+    //   [          value1          ]
+    //                 [          value2          ]
+    //
+    UINT64   value2     = MAKEINT64(cns[len - 4], cns[len - 3], cns[len - 2], cns[len - 1]);
+    GenTree* firstIndir = impCreateCompareInd(this, data, TYP_LONG, dataOffset, value1);
+
+    ssize_t  offset      = dataOffset + len * sizeof(WCHAR) - sizeof(UINT64);
+    GenTree* secondIndir = impCreateCompareInd(this, gtClone(data), TYP_LONG, offset, value2);
+
+    // TODO-Unroll-CQ: Consider merging two indirs via XOR instead of QMARK
+    GenTreeColon* doubleIndirColon = gtNewColonNode(TYP_INT, secondIndir, gtNewFalse());
+    return gtNewQmarkNode(TYP_INT, firstIndir, doubleIndirColon);
+#else // TARGET_64BIT
+    return nullptr;
+#endif
+}
+
+//------------------------------------------------------------------------
+// impExpandHalfConstEquals: Attempts to unroll and vectorize
+//    Equals against a constant WCHAR data for Length in [8..32] range
+//    using either SWAR or SIMD. In a general case it will look like this:
+//
+//    bool equals = obj != null && obj.Length == len && (SWAR or SIMD)
+//
+// Arguments:
+//    data         - Pointer to a data to vectorize
+//    lengthFld    - Pointer to Length field
+//    checkForNull - Check data for null
+//    startsWith   - Is it StartsWith or Equals?
+//    cns          - Constant data (array of 2-byte chars)
+//    len          - Number of 2-byte chars in the cns
+//    dataOffset   - Offset for data
+//
+// Return Value:
+//    A pointer to the newly created SIMD node or nullptr if unrolling is not
+//    possible or not profitable
+//
+GenTree* Compiler::impExpandHalfConstEquals(
+    GenTree* data, GenTree* lengthFld, bool checkForNull, bool startsWith, WCHAR* cnsData, int len, int dataOffset)
+{
+    assert(len >= 0);
+
+    if (compCurBB->isRunRarely())
+    {
+        // Not profitable to expand
+        JITDUMP("impExpandHalfConstEquals: block is cold - not profitable to expand.\n");
+        return nullptr;
+    }
+
+    if (fgBBcount > 20)
```

Comment:
Does this work right when inlining? Is `fgBBCount` the total of all blocks or just the blocks for the currently imported method?

Reply:
Changed to take inlining into account